frame/support/src/traits/hooks.rs

```diff
@@ -152,11 +152,32 @@ pub trait OnRuntimeUpgradeHelpersExt {
 #[cfg(feature = "try-runtime")]
 impl<U: OnRuntimeUpgrade> OnRuntimeUpgradeHelpersExt for U {}
 
+#[derive(
+	codec::Encode, codec::Decode, Clone, sp_std::fmt::Debug, scale_info::TypeInfo, PartialEq, Eq,
+)]
+pub struct RuntimeUpgradeIdentifier(pub [u8; 16]);
+
+impl From<&str> for RuntimeUpgradeIdentifier {
+	fn from(s: &str) -> Self {
+		let mut dest: [u8; 16] = Default::default();
+		if s.len() > 16 {
+			dest.copy_from_slice(&s.as_bytes()[..16]);
+		} else if s.len() < 16 {
+			((&mut dest)[..s.len()]).copy_from_slice(s.as_bytes());
+		} else {
+			dest.copy_from_slice(s.as_bytes());
+		}
+		RuntimeUpgradeIdentifier(dest)
+	}
+}
+
```

Comment:
A recursive name struct can be added here which is needed for the tuples:  
```rust
#[derive(
	codec::Encode, codec::Decode, Clone, sp_std::fmt::Debug, scale_info::TypeInfo, PartialEq, Eq,
)]
pub enum RuntimeUpgradeName {
	Unnamed,
	Ident(Box<RuntimeUpgradeName>),
	Multi(Vec<Box<RuntimeUpgradeName>>),
}
```

then a `name` function for the `OnRuntimeUpgrade`:  
```rust
fn name() -> RuntimeUpgradeName {
	RuntimeUpgradeName::Unnamed
}
```
and for the tuple:  
```rust
fn name() -> RuntimeUpgradeName {
	let names = [ for_tuples!( #( Box::new(Tuple::name()) ),* ) ];
	RuntimeUpgradeName::Multi(names.to_vec())
}
```

Not sure where the event logic needs to be, but you said it should not reside in the  `OnRuntimeUpgrade`.